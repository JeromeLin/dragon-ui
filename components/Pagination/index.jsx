import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

class Pagination extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: nextProps.value
      });
    }
  }

  render() {
    const { props } = this;
    const {
      isBordered,
      isRadius,
      className,
      total,
      pageSize,
      addonBefore,
      addonAfter,
      style
    } = props;

    const cls = classnames({
      'ui-pagination': true,
      bordered: 'bordered' in props || isBordered,
      radius: 'radius' in props || isRadius,
      [className]: !!className
    });

    const pageCount = Math.ceil(total / pageSize);
    let pagerList = [];
    let { value } = this.state;
    value = value < 1 ? 1 : value;
    value = value > pageCount ? pageCount : value;

    if (pageCount <= 9) {
      for (let i = 1; i <= pageCount; i++) {
        pagerList.push(
          <li
            key={i}
            title={i}
            className={classnames({
              'ui-pagination-item': true,
              'ui-pagination-item-active': value === i
            })}
            onClick={() => this._onPagerClick(i)}
          >
            {i}
          </li>
        );
      }
    } else {
      const firstPager = (
        <li
          key={1}
          title="第一页"
          className="ui-pagination-item"
          onClick={() => this._onPagerClick(1)}
        >
          1
        </li>
      );
      const lastPager = (
        <li
          key={pageCount}
          title="最后一页"
          className="ui-pagination-item"
          onClick={() => this._onPagerClick(pageCount)}
        >
          {pageCount}
        </li>
      );
      const prevPager = (
        <li
          key="prev"
          title="上一页"
          className={classnames({
            'ui-pagination-item': true,
            'ui-pagination-item-prev': true,
            'ui-pagination-item-disabled': Number(value) === 1
          })}
          onClick={() => value > 1 && this._onPagerClick(value - 1)}
        >
          <Icon type="arrow-left" />
        </li>
      );
      const nextPager = (
        <li
          key="next"
          title="下一页"
          className={classnames({
            'ui-pagination-item': true,
            'ui-pagination-item-next': true,
            'ui-pagination-item-disabled': Number(value) === pageCount
          })}
          onClick={() => value < pageCount && this._onPagerClick(value + 1)}
        >
          <Icon type="arrow-right" />
        </li>
      );
      const jumpPrev = (
        <li
          key="jump-prev"
          title="向前 5 页"
          className="ui-pagination-item ui-pagination-item-jump-prev"
          onClick={() => this._onPagerClick(value - 5)}
        />
      );
      const jumpNext = (
        <li
          key="jump-next"
          title="向后 5 页"
          className="ui-pagination-item ui-pagination-item-jump-next"
          onClick={() => this._onPagerClick(value + 5)}
        />
      );

      let left = Math.max(1, value - 2);
      let right = Math.min(value + 2, pageCount);

      if (value - 1 <= 2) {
        right = 1 + 4;
      }

      if (pageCount - value <= 2) {
        left = pageCount - 4;
      }

      for (let i = left; i <= right; i++) {
        pagerList.push(
          <li
            key={i}
            title={i}
            className={classnames({
              'ui-pagination-item': true,
              'ui-pagination-item-active': value === i
            })}
            onClick={() => this._onPagerClick(i)}
          >
            {i}
          </li>
        );
      }

      if (value - 1 >= 4) {
        pagerList.unshift(jumpPrev);
      }
      if (pageCount - value >= 4) {
        pagerList.push(jumpNext);
      }

      if (left !== 1) {
        pagerList.unshift(firstPager);
      }
      if (right !== pageCount) {
        pagerList.push(lastPager);
      }

      pagerList.unshift(prevPager);
      pagerList.push(nextPager);
    }

    if (addonBefore) {
      pagerList.unshift(
        <div key="addon-before" className="ui-pagination-addon-before">
          {addonBefore}
        </div>
      );
    }

    if (addonAfter) {
      pagerList.push(
        <div key="addon-after" className="ui-pagination-addon-after">
          {addonAfter}
        </div>
      );
    }

    return (
      <ul className={cls} style={style}>
        {pagerList}
      </ul>
    );
  }

  _onPagerClick(value) {
    this.setState({
      value
    });
    this.props.onPageChange(value);
  }
}

Pagination.propTypes = {
  defaultValue: PropTypes.number,
  isBordered: PropTypes.bool,
  isRadius: PropTypes.bool,
  total: PropTypes.number,
  pageSize: PropTypes.number,
  onPageChange: PropTypes.func
};

Pagination.defaultProps = {
  defaultValue: 1,
  isBordered: false,
  isRadius: false,
  total: 0,
  pageSize: 10,
  onPageChange: () => {}
};

export default Pagination;
