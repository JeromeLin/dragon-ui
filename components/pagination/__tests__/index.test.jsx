import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Pagination from '../index';

describe('Pagination', () => {
  it('renders normal Pagination correctly', () => {
    const wrapper = render(
      <div>
        <Pagination total={100} />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Pagination with total info correctly', () => {
    const wrapper = render(
      <div>
        <Pagination total={100} showTotal />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Pagination with jumper correctly', () => {
    const wrapper = render(
      <div>
        <Pagination total={100} showJumper />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Pagination with customized props correctly', () => {
    const wrapper = render(
      <div>
        <Pagination
          total={100}
          addonBefore="addonBefore"
          addonAfter="addonAfter"
        />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when changing page', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Pagination total={100} onPageChange={onChange} />
      </div>,
    );

    wrapper
      .find('.zw-pagination-item')
      .at(2)
      .simulate('click');

    expect(onChange).toHaveBeenCalledWith(2);
  });

  it('behaves correctly when click prev button', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Pagination total={100} value={6} onPageChange={onChange} />
      </div>,
    );

    wrapper.find('.zw-pagination-item-prev').simulate('click');

    expect(onChange).toHaveBeenCalledWith(5);
  });

  it('behaves correctly when click next button', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Pagination total={100} value={5} onPageChange={onChange} />
      </div>,
    );

    wrapper.find('.zw-pagination-next').simulate('click');

    expect(onChange).toHaveBeenCalledWith(6);
  });

  it('behaves correctly when click prev 5 button', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Pagination total={100} value={6} onPageChange={onChange} />
      </div>,
    );

    wrapper.find('.zw-pagination-item-jump-prev').simulate('click');

    expect(onChange).toHaveBeenCalledWith(1);
  });

  it('behaves correctly when click next 5 button', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Pagination total={100} value={1} onPageChange={onChange} />
      </div>,
    );

    wrapper.find('.zw-pagination-item-jump-next').simulate('click');

    expect(onChange).toHaveBeenCalledWith(6);
  });

  it('behaves correctly when change page with jumper', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Pagination showJumper total={100} value={1} onPageChange={onChange} />
      </div>,
    );

    wrapper
      .find('.zw-pagination-options-quick-jumper input')
      .simulate('keydown', {
        keyCode: 13,
        target: {
          value: 6,
        },
      });
    expect(onChange).toHaveBeenCalledWith(6);
  });

  it('behaves correctly when receiving new value', () => {
    const wrapper = mount(<Pagination showJumper total={100} value={2} />);

    wrapper.setProps({ value: 1 });
    expect(
      wrapper
        .find('.zw-pagination-item')
        .at(1)
        .hasClass('zw-pagination-item-active'),
    ).toBeTruthy();
  });

  it('renders size correctly', () => {
    const wrapper = render(
      <div>
        <Pagination total={100} size="small" />
      </div>,
    );
    expect(wrapper.find('.zw-pagination-small').length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
