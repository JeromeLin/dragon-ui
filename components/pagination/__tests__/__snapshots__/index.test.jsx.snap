// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders Pagination with customized props correctly 1`] = `
<div>
  <ul
    class="zw-pagination"
  >
    <li
      class="zw-pagination-addon-before"
    >
      addonBefore
    </li>
    <li
      class="zw-pagination-item zw-pagination-prev zw-pagination-item--disabled"
      title="上一页"
    >
      <i
        class="zw-icon zw-icon--arrow-left zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-left"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-left"
          />
        </svg>
      </i>
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--active"
      title="1"
    >
      1
    </li>
    <li
      class="zw-pagination-item"
      title="2"
    >
      2
    </li>
    <li
      class="zw-pagination-item"
      title="3"
    >
      3
    </li>
    <li
      class="zw-pagination-item"
      title="4"
    >
      4
    </li>
    <li
      class="zw-pagination-item"
      title="5"
    >
      5
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--ellipsis"
      title="向后 5 页"
    />
    <li
      class="zw-pagination-item"
      title="最后一页"
    >
      10
    </li>
    <li
      class="zw-pagination-item zw-pagination-next"
      title="下一页"
    >
      <i
        class="zw-icon zw-icon--arrow-right zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-right"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-right"
          />
        </svg>
      </i>
    </li>
    <li
      class="zw-pagination-addon-after"
    >
      addonAfter
    </li>
  </ul>
</div>
`;

exports[`Pagination renders Pagination with jumper correctly 1`] = `
<div>
  <ul
    class="zw-pagination"
  >
    <li
      class="zw-pagination-item zw-pagination-prev zw-pagination-item--disabled"
      title="上一页"
    >
      <i
        class="zw-icon zw-icon--arrow-left zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-left"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-left"
          />
        </svg>
      </i>
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--active"
      title="1"
    >
      1
    </li>
    <li
      class="zw-pagination-item"
      title="2"
    >
      2
    </li>
    <li
      class="zw-pagination-item"
      title="3"
    >
      3
    </li>
    <li
      class="zw-pagination-item"
      title="4"
    >
      4
    </li>
    <li
      class="zw-pagination-item"
      title="5"
    >
      5
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--ellipsis"
      title="向后 5 页"
    />
    <li
      class="zw-pagination-item"
      title="最后一页"
    >
      10
    </li>
    <li
      class="zw-pagination-item zw-pagination-next"
      title="下一页"
    >
      <i
        class="zw-icon zw-icon--arrow-right zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-right"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-right"
          />
        </svg>
      </i>
    </li>
  </ul>
</div>
`;

exports[`Pagination renders Pagination with total info correctly 1`] = `
<div>
  <ul
    class="zw-pagination"
  >
    <li
      class="zw-pagination-total"
    >
      共有 100 条记录 第 1 / 10 页
    </li>
    <li
      class="zw-pagination-item zw-pagination-prev zw-pagination-item--disabled"
      title="上一页"
    >
      <i
        class="zw-icon zw-icon--arrow-left zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-left"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-left"
          />
        </svg>
      </i>
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--active"
      title="1"
    >
      1
    </li>
    <li
      class="zw-pagination-item"
      title="2"
    >
      2
    </li>
    <li
      class="zw-pagination-item"
      title="3"
    >
      3
    </li>
    <li
      class="zw-pagination-item"
      title="4"
    >
      4
    </li>
    <li
      class="zw-pagination-item"
      title="5"
    >
      5
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--ellipsis"
      title="向后 5 页"
    />
    <li
      class="zw-pagination-item"
      title="最后一页"
    >
      10
    </li>
    <li
      class="zw-pagination-item zw-pagination-next"
      title="下一页"
    >
      <i
        class="zw-icon zw-icon--arrow-right zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-right"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-right"
          />
        </svg>
      </i>
    </li>
  </ul>
</div>
`;

exports[`Pagination renders normal Pagination correctly 1`] = `
<div>
  <ul
    class="zw-pagination"
  >
    <li
      class="zw-pagination-item zw-pagination-prev zw-pagination-item--disabled"
      title="上一页"
    >
      <i
        class="zw-icon zw-icon--arrow-left zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-left"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-left"
          />
        </svg>
      </i>
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--active"
      title="1"
    >
      1
    </li>
    <li
      class="zw-pagination-item"
      title="2"
    >
      2
    </li>
    <li
      class="zw-pagination-item"
      title="3"
    >
      3
    </li>
    <li
      class="zw-pagination-item"
      title="4"
    >
      4
    </li>
    <li
      class="zw-pagination-item"
      title="5"
    >
      5
    </li>
    <li
      class="zw-pagination-item zw-pagination-item--ellipsis"
      title="向后 5 页"
    />
    <li
      class="zw-pagination-item"
      title="最后一页"
    >
      10
    </li>
    <li
      class="zw-pagination-item zw-pagination-next"
      title="下一页"
    >
      <i
        class="zw-icon zw-icon--arrow-right zw-icon--default zw-icon--md"
        theme="default"
        type="arrow-right"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <use
            href="#arrow-right"
          />
        </svg>
      </i>
    </li>
  </ul>
</div>
`;
